learning/tour-of-beam/frontend/lib/pages/tour/state.dart

```diff
@@ -178,114 +180,82 @@ class TourNotifier extends ChangeNotifier with PageStateMixin<void> {
         snippetEditingController.activeFileController?.isChanged ?? false;
     final snippetFiles = snippetEditingController.getFiles();
 
-    final doSave = _authNotifier.isAuthenticated &&
-        isCodeChanged &&
-        currentUnitController != null &&
-        snippetFiles.isNotEmpty;
+    final doSave =
+        isCodeChanged && _currentUnitContent != null && snippetFiles.isNotEmpty;
 
     if (doSave) {
       _saveCodeDebounced?.call([], {
-        const Symbol('sdkId'): currentUnitController!.sdkId,
-        const Symbol('unitId'): currentUnitController!.unitId,
+        const Symbol('sdk'): currentSdk,
+        const Symbol('unitId'): currentUnitId,
         const Symbol('snippetFiles'): snippetFiles,
       });
     }
   }
 
-  Future<void> _saveUserCode({
-    required String sdkId,
+  Future<void> _saveCode({
+    required Sdk sdk,
     required List<SnippetFile> snippetFiles,
     required String unitId,
   }) async {
     saveCodeStatus = SaveCodeStatus.saving;
     try {
-      final client = GetIt.instance.get<TobClient>();
-      await client.postUserCode(
-        sdkId: sdkId,
+      await _unitProgressCache.saveSnippet(
+        sdk: sdk,
         snippetFiles: snippetFiles,
         unitId: unitId,
       );
-      await _updateHasSavedSnippet();
       saveCodeStatus = SaveCodeStatus.saved;
+      await _updateSnippetTypeSelector();
+      _trySetSnippetType(SnippetType.saved);
     } on Exception catch (e) {
       print(['Could not save code: ', e]);
       _saveCodeStatus = SaveCodeStatus.error;
     }
   }
 
-  Future<void> showSnippetByType(SnippetType snippetType) async {
-    if (snippetType == _snippetType) {
-      return;
-    }
-    _snippetType = snippetType;
-    // Get fresh saved snippet IDs.
+  Future<void> _updateSnippetTypeSelector() async {
```

Comment:
Inline this function to all locations where it is called. Check that `notifyListners` is called properly.